### ISSUE TITLE
self.state:  weakly-referenced object no longer exists

### ISSUE THREAD

**Issue by m1ghtym0** [angr/angr#659]
```
angr/angr/state_plugins/history.py in `reachable(self):
    276             pass
    277         elif self.state is not None:
--> 278             self._satisfiable = self.state.se.satisfiable()
    279         else:
    280             solver = claripy.Solver()
ReferenceError: weakly-referenced object no longer exists
```
I just stumbled across this error, which happens during an exploration step.

**Comment by rhelmot** [comment]
Just normal exploration? It just randomly happened during simgr.explore with nothing else noteworthy applied?

**Comment by ltfish** [comment]
A side note: I also saw it before.

**Comment by m1ghtym0** [comment]
@rhelmot  Sorry for the short description. I use the DFS, Explorer and my own technique, the latter is just for ensuring that I don't run out of memory. I'm not sure what else could be important, but I will try to craft a basic example to showcase this

**Comment by m1ghtym0** [comment]
The error occurs when an SimUnsatError occurs during stepping:
```
       except (SimUnsatError, claripy.UnsatError) as e:
            new_stashes = self._make_stashes_dict(pruned=[a])
            if self._hierarchy:
                self._hierarchy.unreachable_state(a) <----- here
                self._hierarchy.simplify()
```

It all boils down to this:
```
angr/angr/state_hierarchy.py(143)_find_root_unreachable()
    141     def _find_root_unreachable(self, h):
    142         lineage = self.lineage(h)
--> 143         if len(lineage) == 0 or lineage[-1]().reachable():
    144             return h
    145 
```
Where h is `<weakref at 0x7fc2a42d3050; to 'SimStateHistory' at 0x7fc2a431df10>`
and h().state `<SimState @ 0x4000038>`.
But lineage[-1] is `<weakref at 0x7fc2a4636aa0; to 'SimStateHistory' at 0x7fc2a4a93210>`
and lineage[-1]().state `<weakproxy at 0x7fc2a4643838 to NoneType at 0x7fc2bd41bba0>`

**Comment by likebreath** [comment]
Couple of actionable test cases to reproduce this problem can be found #672. I hope it helps and look forward to updates on this issue.

**Comment by zardus** [comment]
@Lukas-Dresel, so we're on the same page, are you planning on looking into this or should I? Either way works for me!

**Comment by m1ghtym0** [comment]
This bug is gone for me!:-)
@likebreath @ltfish What about you guys? Can we close this issue?

**Comment by ltfish** [comment]
Do we know why it's gone?

**Comment by likebreath** [comment]
I still can reproduce this error with the latest version of angr (026382414b7f0445f4648bb6ff915c2f81769e19) being setup by [angr-dev](https://github.com/angr/angr-dev/commit/d1f47e106f30ea78af1ed31f65fb1576ecaef99a). The test case I used is from issue #672.

**Comment by ltfish** [comment]
Yeah I don't think this issue will magically go away until we specifically do something about it...

**Comment by cristipurdel** [comment]
Not sure if it is connected, but I am still getting this error if I use following command to flush the gpu memory
   
`cuda.current_context().reset()`

```
return self.module.context.device

ReferenceError: weakly-referenced object no longer exists
```

**Comment by cgenta20** [comment]
This can be solved by freezing and unfreezing garbage collection in your custom plugins copy function like so:
``` python
@angr.SimStatePlugin.memo
    def copy(self, memo):
        gc.freeze()
        o = super().copy(memo)
        o.field1 = self.field1
        ...
        gc.unfreeze()
        return o
```
